rfcs/0008-transition-frontier-controller.md

```diff
@@ -0,0 +1,337 @@
+populate more information in our transition-frontier-tree and completing the
+add. It is also responsible for handling sync-ledger answers that catchup
+jobs started by other nodes trigger via sending queries.
+
+The expected outcome is to have something with feature parity to the existing
+ledger-builder-controller from the outside. Internally, aside from having
+rigor around the async control flow, we also will take advantage of merkle
+masks to remove a large chunk of complex logic from the existing
+ledger-builder-controller code.
+
+It should also be much easier to unit-test and trace the flow of information
+for debugging purposes.
+
+## Detailed design
+
+[detailed-design]: #detailed-design
+
+### Introduction
+
+`Ledger_builder_controller` was far too big in its scope. In this rewrite,
+we're not only breaking apart the code into more modules, but also breaking it
+up across several libraries.
+
+There will be a new top-level `Transition_frontier_controller` module that will
+just wire the pipes together for the new components:
+
+* [Transition Handler](#transition-handler)
+* [Transition Frontier](#transition-frontier)
+* [Ledger Catchup](#ledger-catchup)
+* [Sync Handler](#sync-handler)
+
+Note that all components communicate between each other solely through the use
+of pipes. By solely using pipes to connect components we can be more explicit about the [Async Control Flow](#async-control-flow).
+
+After describing all the components, we'll cover the specific behavior of
+the [Async Control Flow](#async-control-flow) holistically.
+
+Before covering any of the components it's worth going over some fundamental
+changes in behavior and structure in more detail in [big Changes](#big-changes)
+
+<a href="big-changes"></a>
+### Big Changes
+
+Major differences in behavior between this component and the existing ledger-
+builder-controller are as follows:
+
+1. We only sync-ledger to the locked-ledger (if necessary) and do a new
+[History sync](#history-sync) for catchup jobs
+
+Rationale: Without history sync, we are vulnerable to believing an invalid
+state: Certain parts of the ledger-builder state are not checked in a SNARK and
+so we must get enough info to go back to the locked state (where we know we've
+achieved consensus).
+
+2. We take advantage of the merkle-mask to instead of transitions, putting a `Breadcrumb.t` which contains an `External_transition.t` and a now light-weight `Ledger_builder.t` it also has a notion of the prior breadcrumb. See [RFC-0007](0007-persistent-ledger-builder-controller) for more info on merkle masks, and [Transition frontier](#transition-frontier) for more info about how these masks are configured.
+
+Rationale: We no longer need the asynchronous `Path_traversal` job that is present in the current ledger-builder-controller (simplifying complexity), and the frequent operation of answering sync-ledger queries no longer require materializing ledger-builders.
```

Comment:
This is not the point that allows us to avoid `Path_traversal`. We can avoid `Path_traversal` by using the lookup table. This point allows us to avoid intermediate materialization of nodes.